File: src/bumperbot_py_examples/bumperbot_py_examples/simple_parameter.py
import rclpy
from rclpy.node import Node
from rcl_interfaces.msg import SetParametersResult
from rclpy.parameter import Parameter


class SimpleParameter(Node):
  def __init__(self):
    super().__init__("simple_parameter")

    self.declare_parameter("simple_int_param", 28)
    self.declare_parameter("simple_string_param", "Mark")

    self.add_on_set_parameters_callback(self.paramChangeCallback)

  def paramChangeCallback(self, params):
    result = SetParametersResult()

    for param in params:
      if param.name == "simple_int_param" and param.type == Parameter.Type.INTEGER:
        self.get_logger().info("Param simple_int_param changed: new %d" % param.value)
        result.successful = True
      elif param.name == "simple_string_param" and param.type == Parameter.Type.STRING:
        self.get_logger().info("Param simple_string_param changed: new %s" % param.value)
        result.successful = True
      else:
        self.get_logger().error("Param %s is invalid" % param.name)
        result.successful = False

    return result
    
def main(): 
  rclpy.init()
  simple_parameter = SimpleParameter()
  rclpy.spin(simple_parameter)
  simple_parameter.destroy_node()
  rclpy.shutdown()

if __name__ == "__main__":
  main()

File: src/bumperbot_cpp_examples/src/simple_parameter.cpp
#include <rclcpp/rclcpp.hpp>
#include <rcl_interfaces/msg/set_parameters_result.hpp>

#include <string>
#include <vector>
#include <memory>

using std::placeholders::_1;

class SimpleParameter: public rclcpp::Node {
  public:
    SimpleParameter() : Node("simple_parameter") {
      declare_parameter<int>("simple_int_param", 22);
      declare_parameter<std::string>("simple_string_param", "Maddog");

      param_callback_handle_ = add_on_set_parameters_callback(std::bind(&SimpleParameter::paramChangeCallback, this, _1));
    }

  private:
    OnSetParametersCallbackHandle::SharedPtr param_callback_handle_;

    rcl_interfaces::msg::SetParametersResult paramChangeCallback(const std::vector<rclcpp::Parameter> & params) {
      rcl_interfaces::msg::SetParametersResult result;

      for (const auto& param : params) {
        if (param.get_name() == "simple_int_pram" && param.get_type() == rclcpp::ParameterType::PARAMETER_INTEGER) {
          RCLCPP_INFO_STREAM(get_logger(), "Param simple_int_param changed: " << param.as_int());
          result.successful = true;
        } else if (param.get_name() == "simple_string_param" && param.get_type() == rclcpp::ParameterType::PARAMETER_STRING) {
          RCLCPP_INFO_STREAM(get_logger(), "Param simple_string_param changed: " << param.as_string());
          result.successful = true;
        } else {
          RCLCPP_ERROR_STREAM(get_logger(), "Param " << param.get_name() << " does not exist!!!");
          result.successful = false;
        }
      }

      return result;
    }
};

int main(int argc, char* argv[]) {
  rclcpp::init(argc, argv);

  auto node = std::make_shared<SimpleParameter>();
  rclcpp::spin(node);
  rclcpp::shutdown();


  return 0;
}